src/main/daml/Daml/Finance/Holding/Account.daml

```diff
@@ -34,26 +38,47 @@ template Account
     signatory custodian, owner
     observer Disclosure.flattenObservers observers
 
-    let accountKey = AccountKey with custodian; owner; id
+    let
+      account = AccountKey with custodian; owner; id
+      transferApprovers = S.singleton owner
 
     interface instance Account.I for Account where
       asDisclosure = toInterface @Disclosure.I this
-      view = Account.View with custodian; id; owner; holdingFactoryCid; description
-      getKey = accountKey
+      view = Account.View with custodian; id; owner; holdingFactoryCid; description; transferApprovers
+      getKey = account
       credit Account.Credit{quantity} =
         exercise holdingFactoryCid HoldingFactory.Create with
           instrument = quantity.unit
-          account = accountKey
+          account
           amount = quantity.amount
           observers = M.empty
-      debit Account.Debit{holdingCid} =
-        exercise holdingFactoryCid HoldingFactory.Remove with actors = fromList [accountKey.custodian, accountKey.owner]; holdingCid
+      debit Account.Debit{holdingCid} = do
+        vHolding <- view <$> fetch holdingCid
+        assertMsg "accounts must match" $ vHolding.account == account
+        exercise holdingFactoryCid HoldingFactory.Remove with actors = S.fromList [account.custodian, account.owner]; holdingCid
+      creditAndDebit Account.CreditAndDebit{debitTransferableCid; actors} = do
+        let lockableCid = toInterfaceContractId @Lockable.I debitTransferableCid
+        lockable <- fetch lockableCid
+        let
+          vH = view $ toInterface @Base.I lockable
+          vL = view $ toInterface @Lockable.I lockable
+        -- Release all locks
+        debitTransferableCid <- toInterfaceContractId @Base.I <$> optional (pure lockableCid) (releaseAll lockableCid) vL.lock
+        -- Debit holding
+        Account.exerciseInterfaceByKey @Account.I vH.account vH.account.owner Account.Debit with holdingCid = debitTransferableCid
```

Comment:
Is there a specific reason we debit the sender holding in this choice here, and not right away in `transferImpl`? I wonder if there is privacy leakage with the debit being part of this exercise / sub-tx, such that the receiver gets to see / witness the senders holding and account. But thinking about it, the receiver is anyway (usually) a controller on the `Transfer` so would get to see everything involved anyway. But would they also if it's a custodian-only-controlled transfer?

Reply:
the `releaseLocks` and `debit` part could indeed be executed in the `transferImpl`, but the `reapplyLocks` and `credit` part must be there. I kept them together to make sure that:
 
1. the debit and credit are executed atomically in the `CreditAndDebit` choice, and the custodian can be certain that the quantity is preserved. if we split them up, we would probably need to protect the choice by a call token or something (to preserve quantities).
2. to avoid having to communicate the `lock` to reapply (from the `transferImpl` to the `CreditAndDebit` impl). 

but from a privacy perspective it is indeed better to keep them apart. for the custodian-only-controlled case, neither the sender nor the receiver would learn to know of each other which is pretty cool.

Reply:
Ah, I see that we need to re-apply the locks, so we need to do both in here (unless we somehow pass those along).

Reply:
thought about the ^ a bit more:

1. the sender will actually always learn to know about the accounts/holdings/credits/debits independent of who triggers it (as the sender is a signatory on the holding triggering the transfer). 
2. if we would split up the debit and credit, we could make it such that the receiver would not learn to know about the sender in the custodian-only-controlled model. 
3. if we leave it as is, it is more symmetric (in the sense that the receiver also learns to know about the sender). it might actually be good for the receiver to see that a transfer credit also debited someone else (and did not happen out of thin air). 
4. in case the privacy for the receiver is a concern, a `Batch` could be created with `sender -> custodian -> receiver`. then the receiver would not necessarily need to know about the sender (and vice versa).